File: public/image/crawl.js
const axios = require('axios')
const fs = require('fs');
const downloader = require('image-downloader');
const mysql = require('mysql');
var productID = 1;

const con = mysql.createConnection({
    host: "localhost",
    user: "root",
    password: "87654321",
    database: "dbtada"
});

con.connect(function (err) {
    if (err) throw err;
    console.log("Connected!");
});

function createFolder(path) {
    try {
        if (!fs.existsSync(path)) {
            fs.mkdirSync(path);
        }
    } catch (err) {
        console.log(err)
    }
}

async function getHTML(url){
    let html = ""
    await axios.get(url)
        .then(datas => {
            html = datas.data;
    });
    return html;
}

async function getDataFromHTML(html, regex) {
    var matchDatas = [];
    let m;
    while ((m = regex.exec(html)) !== null) {
        // This is necessary to avoid infinite loops with zero-width matches
        if (m.index === regex.lastIndex) {
            regex.lastIndex++;
        }
        m.forEach((match, groupIndex) => {
            if (groupIndex == 1) {
                if(matchDatas.indexOf(match) === -1)
                matchDatas.push(match);
            }
        })

    }
    return new Promise((resolve, reject) =>{
        resolve(matchDatas);
    })
}

async function getLinkProduct(html) {
    let result = [];
    const regexLink = /<a target="_blank" href="(.*)" class="img-link-pr" title="(.*)">/gm;
    const regexImage = /<img class="img-responsive lazyload" data-src="(.*)" src=/gm;
    let dataLink = await getDataFromHTML(html, regexLink);
    let dataImage  = await getDataFromHTML(html, regexImage);

    for(var i = 0; i < dataLink.length; i++){
        var tmp =  dataLink[i].split("/");
        var groupName = tmp[tmp.length-1].split("?")[0];
        if(groupName.length > 9) continue;
        if(!dataImage[i]) continue;
        result.push( { link: "https://coupletx.com" + dataLink[i], imageLink: dataImage[i]});
    }
    return result;
}

async function getProductFromLink(html, url, img) {
    var color = url.split("=")[1];
    var tmp = url.split("/");
    var groupName = tmp[tmp.length-1].split("?")[0];
    var sizes ;
    let detail;
    let nameProductVN = [];
    let prices = [];
    let images = [];
    let groupProductID = 0;
    let productByColorID = 0;
    let smallImage = "products/a/" + groupName + "/img"+ parseInt(Math.random()*1000000) + ".jpg";

    let regexName = /<div class="product-title">\n							<h4>(.*)<\/h4>/gm;
    let regexPrice = /<div class="product-price" id="price-preview">\n<span>(.*)<\/span>/gm;
    let regexDetail = /<div class="note-detail" style="height:200px;overflow: scroll;padding:10px">\n								<span style="color:#000;font-size:12px;">\n(.*)<\/span>/gm;
    let regexSize = /<div data-value="(.*)" class="swatch-element">/gm;
    let regexIMG = /<img class="product-image-feature a" src="(.*)" alt="(.*)">/gm;

    await getDataFromHTML(html, regexName)
        .then(data => {
            nameProductVN = data;
        })
        .then(
            getDataFromHTML(html, regexDetail)
            .then(data => {detail = data[0].split("'").join(' ').trim()})
        )
        .then(
            getDataFromHTML(html, regexPrice)
            .then(data => prices = data)
        )
        .then(
            getDataFromHTML(html, regexSize)
            .then(data => sizes = data)
        )
        .then(
            getDataFromHTML(html, regexIMG)
            .then(datas =>{
                datas.forEach(data =>{
                    data = 'https:' + data;
                    var tmp = data.split('/');
                    var imgName = tmp[tmp.length - 1];
                    imgName = imgName.replace('?', '_');
                    const image = {
                        url: data,
                        dest: 'products/a/' + groupName + '/' + imgName
                    }
                    images.push(image);
                });
            })
        )
        .then(() =>{
            inserSQL('INSERT INTO `groupproducts`(`GroupName`, `GroupNaaoVN`, `Price`, `Description`, `Sale`, `CategoryID`, `Type`) VALUES (\''+nameProductVN[0]+'\',\''+groupName+'\',\''+prices[0]+'\',\''+detail+'\',0, 16,\'a\')')
            .then((result) =>{
                createFolder("./products/a/" + groupName);
                console.log("Inserted", groupName);
            })
            .catch(err => {
                console.log("error when insert", groupName, '\n');
                // if(err.message.indexOf("Data too long") != -1) console.log('INSERT INTO `groupproducts`(`GroupName`, `GroupNameNoVN`, `Price`, `Description`, `Sale`, `CategoryID`, `Type`) VALUES (\''+nameProductVN[0]+'\',\''+groupName+'\',\''+prices[0]+'\',\''+detail+'\',0, 1,\'Women\')');
            });
        });

    await selectSQL('SELECT `GroupProductID` FROM `groupproducts` WHERE GroupNameNoVN = \''+groupName+'\'')
        .then(result =>{
            if(result[0]) groupProductID = result[0]['GroupProductID'];
    });

    await inserSQL("INSERT INTO `productbycolors`(`SmallImage`, `Color`, `GroupProductID`) VALUES ('"+smallImage+"', '"+ color+"', "+groupProductID+")")
    .then(() =>{
        downloader.image({
            url: "http:"+img,
            dest: "./"+smallImage
        }).then(console.log("saved small image"))
        .catch(err => console.log("error when save small image", err));
    }).catch(err => console.log("error when insert small image", err));

    await selectSQL('SELECT `ProductByColorID` FROM `productbycolors` WHERE Color = \''+color+'\' and ' + "GroupProductID = "+groupProductID)
        .then(result =>{
            productByColorID = result[0]['ProductByColorID'];
    });

    await images.forEach(function(image){
        inserSQL('INSERT INTO `productimages`(`Path`, `ProductByColorID`) VALUES (\''+image.dest+'\','+productByColorID+')')
        .then((result) =>{
            console.log("Inserted", image.dest);
            downloader.image({url: image.url, dest: "./" + image.dest})
            .then(({ filename, image }) => {
                console.log('Saved to', filename);
            })
        })
    });

    await sizes.forEach(function (size){
        inserSQL('INSERT INTO `products`(`Size`, `QuantityStorage`, `ProductByColorID`, `QuantitySelled`, `View`) VALUES (\''+size+'\',10,'+productByColorID+',0,0)')
        .then(result => console.log("Inserted", size));
    });

    console.log('\n\n--------------------------------------------------------------');
    console.log('------------------'+groupName+' done--------------------------');
}

function inserSQL(query) {
    return new Promise((resolve, reject) =>{
        con.query(query, function (err, result) {
            if (err) {
                reject(err);
            }else{
                resolve(result);
            }
        })
    });
}

function selectSQL(query) {
    return new Promise((resolve, reject) =>{
        con.query(query, function (err, result) {
            if (err) {
                reject(err);
            }else{
                resolve(result);
            }
        })
    });
}

async function main() {
    let html = await getHTML('https://coupletx.com/collections/a-ba-lo-1');
    let productLinks = await getLinkProduct(html);
    productLinks.forEach(async function(data){
        let htmlProduct = await getHTML(data.link);
        await getProductFromLink(htmlProduct, data.link, data.imageLink);
    })

}

main();

File: public/image/crawl.py
import urllib.request
import re
import pymysql
import os
import random
import json
from urllib.request import urlopen, Request

db = pymysql.connect("localhost", "root", "87654321", "AnShop")
cursor = db.cursor()

patterns = {
    '[àáảãạăắằẵặẳâầấậẫẩ]': 'a',
    '[đ]': 'd',
    '[èéẻẽẹêềếểễệ]': 'e',
    '[ìíỉĩị]': 'i',
    '[òóỏõọôồốổỗộơờớởỡợ]': 'o',
    '[ùúủũụưừứửữự]': 'u',
    '[ỳýỷỹỵ]': 'y',
    ' ': '-'
}

def convert(text):
    output = text
    for regex, replace in patterns.items():
        output = re.sub(regex, replace, output)
        # deal with upper case
        output = re.sub(regex.upper(), replace.upper(), output)
    return output.lower()

def downloadIMG(url, path):
    opener=urllib.request.build_opener()
    opener.addheaders=[('User-Agent','Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/36.0.1941.0 Safari/537.36')]
    urllib.request.install_opener(opener)
    urllib.request.urlretrieve(url, path)
    print("\tSave to ", path)

def getHTMLFromUrl(url):
    fp = urllib.request.urlopen(url)
    mybytes = fp.read()
    html = mybytes.decode("utf8")
    fp.close()
    return html


def getLinkProduct(html):
    result = []
    regexLink = r'<a target=\"_blank\" href=\"(.*)\" class=\"img-link-pr\" title=\"(.*)\">'
    regexImage = r'<img class=\"img-responsive lazyload\" data-src=\"(.*)\" src='
    dataLink = re.findall(regexLink, html)
    dataImage  = re.findall(regexImage, html)

    for i in range(dataLink.count):
        tmp =  dataLink[i].split("/");
        groupName = tmp[tmp.length-1].split("?")[0];
        if(groupName.length > 9):
            continue
        if(dataImage[i] is None):
            continue
        result.push( ["https://coupletx.com" + dataLink[i], dataImage[i]])
    return result

def getProduct(url, idCategory, idType):
    html = getHTMLFromUrl(url)

    rawData = json.loads(str(re.findall(r'window.productJSON = (.*);', html)[0]))
    code= rawData["handle"]
    name = rawData["title"]
    description = rawData["description"]
    price = int(rawData["price"])
    featuredImage = rawData["featured_image"]
    images = rawData["images"]
    type = rawData["type"]
    idGroup = -1

    path = "images/products/"+type+"/"+code+"/"
    if not os.path.exists(path):
        os.makedirs(path)
        featuredImageName = featuredImage.split("_")[len(featuredImage.split("_"))-1]
        downloadIMG(featuredImage, path+featuredImageName)
    else:
        print("Product ", name, "already existed")
        return

    query = "INSERT INTO `AnShop`.`group_products` (`group_name`, `group_code`, `price`, `desciption`, `image`, `id_category`, `type`) VALUES (%s, %s, %d, %s, %s, %d, %d)"
    value = [name, code, price, description, path+featuredImageName, idCategory, idType]
    try:
        cursor.execute(query, value)
        db.commit()
        print("Inserted "+name)
        idGroup = int(cursor.lastrowid)
    except:
        print("Error in insert", name)
        return

    if idGroup == -1:
        print("Error when get id group")
        return

    imagePerColor = {}

    for image in images:
        correction = re.findall(r'product/(.*)_(.*)__\d__(.*)', image)
        if len(correction) > 0:
            correction = correction[0]
            if correction[1] not in imagePerColor:
                imagePerColor[str(correction[1])] = []
            imagePerColor[str(correction[1])].append({"image_name": correction[2], "link": image})
        else :
            retry = re.findall(r'product/(.*)_(.*)_(.*)', image)
            if len(retry) > 0:
                retry = retry[0]
                if retry[0] == 'men' or retry[0] == 'lady' or retry[0] == 'couple':
                    if retry[1] not in imagePerColor:
                        imagePerColor[str(retry[1])] = []
                    imagePerColor[str(retry[1])].append({"image_name":retry[2], "link":image})

    colors = []
    idColor = -1
    for item in rawData["variants"]:
        color = item["option1"]
        if color not in colors:
            colors.append(color)
            cursor.execute("INSERT INTO `AnShop`.`product_color` (`color`, `id_group_product`) VALUES ( %s, %d)", [color, idGroup])
            db.commit()
            idColor = int(cursor.lastrowid)
            for image in imagePerColor[convert(color)]:
                try:
                    queryImage = "INSERT INTO `AnShop`.`product_images` (`path`, `id_product_color`) values (%s, %d)" % (path+image['image_name'], idColor)
                    cursor.execute(queryImage)
                    db.commit()
                    print("\tInserted ",image)
                    downloadIMG(image['link'], path+image['image_name'])
                except:
                    print("Error in insert", image['image_name'])
        if idColor != -1:
            query = "INSERT INTO `AnShop`.`products` (`size`, `quantity_stogare`, `id_product_color`) values (%s, %d, %d)" % (item["option3"], item["inventory_quantity"], idColor)
            cursor.execute(queryImage)
            db.commit()
        else:
            print("Error when insert size:", color, item["option3"])


html = getHTMLFromUrl("https://coupletx.com/collections/m-ao-so-mi-nam")
matchs = re.findall(r'<a target=\"_blank\" href=\"(.*)\" class=\"img-link-pr\" title=\"(.*)\">', html)
for match in matchs:
    url = 'https://coupletx.com' + match[0]
    getProduct(url, 1, 1)

